File: Stacks/next_greater_element.py

def nextGreaterElement(nums1, nums2):

    hashMap = {n:i for i,n in enumerate(nums1)}

    res = [-1] * len(nums1)

    stack = [] 
    for num in nums2:

        while stack and stack[-1] < num:
            smaller_num = stack.pop()
            if smaller_num in hashMap:
                res[hashMap[smaller_num]] = num
        stack.append(num)

    return res



n1 = [4,1,2]
n2 = [1,3,4,2]
res = nextGreaterElement(n1, n2)

print(res)



def func(arr):

    stack = []
    res = [-1] * len(arr)

    for i,n in enumerate(arr):

        while stack and arr[stack[-1]] < n:
            index = stack.pop()
            res[index] = n
        stack.append(i)


    return res


arr = [1,2,4,2]
res = func(arr)
print(res)

File: Stacks/next_greater_element.js
function nextGreaterNumber(nums1, nums2) {

    let hashM = {}
    let res = new Array(nums1.length).fill(-1);
    nums1.forEach((element, index) => {
        hashM[element] = index
    });

    let stack = []
    nums2.forEach((num) => {
        while (stack && stack[stack.length - 1] < num) {
            let last_num = stack.pop()
            if (last_num in hashM) {
                res[hashM[last_num]] = num
            }
        }
        stack.push(num)
    })

    return res
}



let n1 = [4, 1, 2]
let n2 = [1, 3, 4, 2]
let res = nextGreaterNumber(n1, n2)
console.log(res);
